feat: add release CI

McPatate opened the pull request:
Ported the CI from llm-ls

SilasMarvin commented:
Wow this is really incredible thank you for adding all this! I am not very familiar with GitHub actions and have learned a lot reading through this.

I'm not 100% sure why we changed `pub` to `pub(crate)` I thought that only ever meant anything for libraries? 

Thank you for adding me as an author, I'm not 100% sure I want my email public though. Can we change it to `silas.marvin@dmarvin.net` that is an alias for me.

McPatate commented:
> Wow this is really incredible thank you for adding all this! I am not very familiar with GitHub actions and have learned a lot reading through this.

There are most likely still some things we can do to cache builds but I don't think it's necessary on a release CI given it should be done seldomly! Glad you learned something 😄 
We could also add a CI to run the test suite on every push (caching would definitely be interesting here).

> I'm not 100% sure why we changed pub to pub(crate) I thought that only ever meant anything for libraries?

There will be warnings on some platforms and it'll cause errors in the build. We could disable the visibility warnings but I think they can be useful at times, so I'd rather set the visibility to `pub(crate)`.
Even though it is a binary, I believe visibility is determined at the module level regardless of what the target output is, lib or bin (don't quote me on that I need to check).

> Thank you for adding me as an author, I'm not 100% sure I want my email public though. Can we change it to silas.marvin@dmarvin.net that is an alias for me.

Sorry about that, changed the commit and force pushed, it shouldn't be accessible anymore.
